[fix] fix scalar behavior for `torch.where`

kshitij12345 opened the pull request:
Reference: #9190 
Fixes: #64161
Fixes https://github.com/pytorch/pytorch/issues/73298

Post this PR:
For scalar-tensor variant, the type-promotion rules apply and the dtype of output is determined after considering the dtypes of `x` and `y` of `torch.where` as per https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc
For scalar-scalar variant, we promote to the scalar with highest kind.

cc @mruberry @rgommers @pmeier @asmeurer @leofang @AnirudhDagar @asi1024 @emcastillo @kmaehashi

kshitij12345 committed:
update docs

kshitij12345 committed:
update test for new behaviour

kshitij12345 committed:
check if windows misaligned address is fixed in CI

kshitij12345 committed:
make flake8 happy :)

kshitij12345 committed:
add comment

kshitij12345 commented:
cc: @vadimkantorov

vadimkantorov commented:
there are also similar type promotion issues with torch.lerp which is a "differentiable" version of torch.where. I think it wasn't prioritised, but I still hope it will be fixed as well :)

kshitij12345 commented:
Can you share the link to that issue :)
Thanks!

vadimkantorov commented:
https://github.com/pytorch/pytorch/issues/57947

kshitij12345 commented:
@mruberry gentle ping :)

kshitij12345 commented:
Gentle ping @mruberry

mruberry commented:
> Gentle ping @mruberry

Sorry I haven't had a chance to respond to this yet, @kshitij12345. I've asked @saketh-are to take a look as part of his study of type promotion. I'll be on pto starting Wednesday so you may need to be patient for a bit longer, though. Sorry about that.

kshitij12345 commented:
Hey Mike, thanks for the heads-up! No hurry. Enjoy the PTO :)

vadimkantorov commented:
Does it also fix this?
```
>>> torch.where(torch.tensor([True]), torch.arange(1, dtype = torch.int16), 1)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: expected scalar type short int but found long int
```

kshitij12345 commented:
Yes
```python
>>> import torch
>>> torch.where(torch.tensor([True]), torch.arange(1, dtype = torch.int16), 1)
tensor([0], dtype=torch.int16)
>>> torch.where(torch.tensor([False]), torch.arange(1, dtype = torch.int16), 1)
tensor([1], dtype=torch.int16)
```

mruberry commented:
Sorry for the delay, @kshitij12345, @saketh-are has been filling sick recently :(

kshitij12345 commented:
No hurry. Thanks for updating :)

saketh-are commented on test/test_torch.py:
```diff
@@ -7613,31 +7612,31 @@ def _test_where_scalar_template(self, device, dtype, exec_fn):
     # below are the valid `TensorDtype` and `ScalarType` combinations.
     def _where_valid_scalar_tensor_combination(self, scalar_type, dtype):
-        if (scalar_type == int and dtype == torch.long):
+        # Integral Tensor and Integral Scalar
```
Would it be clearer if we reorganize these checks as follows?
```
if dtype.is_floating_point:
    return scalar_type in [int, float]
if dtype.is_complex:
    return True
return scalar_type == int
```

kshitij12345 replied:
Sure.

saketh-are commented on torch/_torch_docs.py:
```diff
@@ -10302,12 +10302,6 @@ def merge_dicts(*dicts):
 .. note::
     The tensors :attr:`condition`, :attr:`x`, :attr:`y` must be :ref:`broadcastable <broadcasting-semantics>`.
 
-.. note::
```
It's great that we can get rid of this from the documentation now!

saketh-are commented on test/test_torch.py:
```diff
@@ -7613,31 +7612,31 @@ def _test_where_scalar_template(self, device, dtype, exec_fn):
             for ndims in range(0, 4):
                 shape = self._rand_shape(ndims, min_size=5, max_size=10)
-                for n in range(ndims + 1):
-                    for c in combinations(list(range(ndims)), n):
```
Nice catch, if I understand correctly these loop variables were already unused?

kshitij12345 replied:
Yes.

saketh-are commented:
Hey @kshitij12345, thanks a lot for your patience while I was out sick. This looks great and I left a few minor comments inline.

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

kshitij12345 committed:
address review

kshitij12345 commented:
Hey @saketh-are , 

Thanks for reviewing. Have simplified the valid combination logic. Should be ready for another round :)

saketh-are commented on test/test_torch.py:
```diff
@@ -7644,33 +7643,28 @@ def _test_where_scalar_template(self, device, dtype, exec_fn):
-        if (scalar_type == int and dtype == torch.long):
+        if (dtype.is_floating_point):
+            return scalar_type in (int, float)
+        elif dtype.is_complex:
```
Super tiny nit: if/if would be better here (no elif) since the if path returns

kshitij12345 replied:
Done :)

kshitij12345 committed:
address review

vadimkantorov commented:
Curious, does it also support 1) Python scalar 2) broadcasting as the condition argument? (Python scalar may be not very useful since `torch.where(cond, x, y)` can then be replaced by `x if cond else y`, but maybe useful for some tracing, not sure)

kshitij12345 commented:
> Curious, does it also support 1) Python scalar 2) broadcasting as the condition argument?

1. `condition` has to be a tensor. 
2. `condition` is broadcasted if necessary based on the shape of `x` or `y`.

vadimkantorov commented:
Probably this PR fixes also https://github.com/pytorch/pytorch/issues/64161

kshitij12345 commented:
Gentle Ping @saketh-are @mruberry . This should be ready. Thanks!

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

kshitij12345 committed:
support scalar with higher kind than input tensor

kshitij12345 committed:
update test

kshitij12345 committed:
update map

kshitij12345 committed:
make flake8 happy:)

kshitij12345 commented:
XLA failures are real. (I think it would be ok to skip XLA for now).

PTAL @mruberry @saketh-are

JackCaoG commented:
@kshitij12345 Let me take a quick look at this, if this is something that xla can also we fix we should do that

kshitij12345 commented:
@JackCaoG Thanks for prompt look. Do note the the PR description is stale. The behaviour is different in latest commit. Will update the description tomorrow.

JackCaoG commented:
@kshitij12345 I put my analysis in https://github.com/pytorch/xla/issues/3105#issuecomment-908574827. I think I know how to fix it, but I will wait until this pr being reviewed first.

kshitij12345 commented:
> @kshitij12345 I put my analysis in pytorch/xla#3105 (comment). I think I know how to fix it, but I will wait until this pr being reviewed first.

Your analysis looks correct. Will ping you once this PR is approved.

kshitij12345 commented:
@saketh-are have updated the PR so that the type promotion is supported for `tensor-scalar` variants. PTAL :)

kshitij12345 committed:
where scalar variant opinfo samples

saketh-are commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -355,15 +377,30 @@ std::vector<Tensor> where(const Tensor& condition) {
+  Tensor ret = at::empty(self.sizes(), self.options().dtype(common_dtype));
+
+  auto get_promoted = [&common_dtype](const Tensor& t) {
+    // dynamic casting on CPU is not allowed
```
This is interesting. Am I understanding correctly that it's due to https://github.com/pytorch/pytorch/pull/28427#issuecomment-545687342?

kshitij12345 replied:
That looks like the motivation. 
There is a check in `cpu_kernel` for the same (which fails without this casting here)
https://github.com/pytorch/pytorch/blob/30a7c768d77d870c6194f5b6cb0bbd6fe1211069/aten/src/ATen/native/cpu/Loops.h#L257-L258
(called from here)
https://github.com/pytorch/pytorch/blob/30a7c768d77d870c6194f5b6cb0bbd6fe1211069/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L208

AFAIK, TensorIterator does similar things for CPU and type promotion

cc: @mruberry

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

kshitij12345 commented:
@saketh-are @mruberry JIT tracer failure looks real (so is XLA but @JackCaoG has taken care of it)

```
======================================================================
ERROR [3.530s]: test_where_ops (__main__.TestTEFuser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_jit_fuser_te.py", line 1642, in test_where_ops
    self.assertEqual(ref, t(cond, x, y))
  File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1596, in assertEqual
    super().assertTrue(result, msg=self._get_assert_msg(msg, debug_msg=debug_msg))
  File "/opt/conda/lib/python3.6/unittest/case.py", line 682, in assertTrue
    raise self.failureException(msg)
AssertionError: False is not true : Tensors failed to compare as equal!With rtol=1e-07 and atol=1e-07, found 2 element(s) (out of 2) whose difference(s) exceeded the margin of error (including 0 nan comparisons). The greatest difference was 0.0008750000000001812 (3.1415 vs. 3.140625), which occurred at index 0.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test_jit_fuser_te.py", line 1646, in test_where_ops
    " ".join(["Failed:", str(dtype), op.__name__, device])
RuntimeError: Failed: torch.float16 <lambda> cuda
```

Looking at the test,
https://github.com/pytorch/pytorch/blob/8535418a06d75025541370cc656a8b6a0330ca0d/test/test_jit_fuser_te.py#L1616-L1624

It fails at first lambda (maybe fuser has different type promotion code for scalar case)

Should we skip the test for now or ping the fuser team?

Thanks!

saketh-are commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -355,15 +377,30 @@ std::vector<Tensor> where(const Tensor& condition) {
+
+  auto get_promoted = [&common_dtype](const Tensor& t) {
+    // dynamic casting on CPU is not allowed
+    return (t.device().is_cpu()) && !(t.scalar_type() == common_dtype)
```
If we're only promoting here on CPU, have you thought about what happens on CUDA? It seems that unpromoted self/other tensors would make it to the TensorIteratorConfig.

kshitij12345 replied:
Even TensorIterator does not create temporary of `common_dtype` for GPU (since dynamic casting is supported on CUDA)

https://github.com/pytorch/pytorch/blob/f69cf3cf2ff392d7791b7efbc069a66494492a77/aten/src/ATen/TensorIterator.cpp#L410-L449

cc: @mruberry (to correct if I am wrong in my understanding)

saketh-are commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -355,15 +377,30 @@ std::vector<Tensor> where(const Tensor& condition) {
+  // For `where` we only want to consider `self` and `other`,
+  // however right now TensorIterator API doesn't have a way to specify which inputs
+  // should take part in type promotion.
+  auto common_dtype = at::result_type(self, other);
```
`at::result_type` is also implemented for (scalar, tensor) and (tensor, scalar) args. Would it make more sense to make use of that rather than using `wrapped_scalar_tensor_default_dtype` (which is a bit convoluted and also seems to have some comment saying not to use it)?

kshitij12345 replied:
In that case we will need to add new overload for `where` kernel which takes a scalar as parameter (because the current version expects all inputs to be passed via TensorIterator which only takes Tensor inputs)  and add other overload for `_s_where` and corresponding entry in `derivatives.yaml`

https://github.com/pytorch/pytorch/blob/f69cf3cf2ff392d7791b7efbc069a66494492a77/aten/src/ATen/native/cpu/TensorCompareKernel.cpp#L204-L205

saketh-are reviewed (commented):
Hey @kshitij12345, thanks for updating this PR to be more general! I left a couple of comments from discussion with @mruberry.

We were also wondering if you've thought about what happens in the out version if the provided `out` tensor has a "lesser" type than the chosen common dtype for the inputs.

Look forward to hearing your thoughts.

kshitij12345 commented:
> We were also wondering if you've thought about what happens in the out version if the provided out tensor has a "lesser" type than the chosen common dtype for the inputs.

I haven't considered that as there is no out version currently and I am not sure if we would add that in the future. The reason being that `out` shape for `where(Tensor condition) -> Tensor[]` overload is not static (and I feel it would be odd if `out` support is not consistent)

https://github.com/pytorch/pytorch/blob/f69cf3cf2ff392d7791b7efbc069a66494492a77/aten/src/ATen/native/native_functions.yaml#L4542-L4562

Note: NumPy also doesn't have `out` support for ` where`.

kshitij12345 commented:
@mruberry 

Regarding our offline conversation about moving the type-promotion code to `where` instead of `_s_where`. I think that would be tricky. For eg. suppose `x` is `float` and `y` is `int64`. Our output should be float (which is allocated by `_s_where`. If these tensors reside on CUDA, then we don't need to cast these.
So for `_s_where` to know the `dtype` of output, it will have to check which input (passed by `where`) has higher dtype and allocate output accordingly (for which we will use `promote_types`. So we are sorta doing promotion computation again) 

https://github.com/pytorch/pytorch/blob/64d3c7388f3ebb084c6e6998597e94753fde1ffa/aten/src/ATen/native/TensorCompare.cpp#L335

https://github.com/pytorch/pytorch/blob/64d3c7388f3ebb084c6e6998597e94753fde1ffa/aten/src/ATen/native/TensorCompare.cpp#L359-L360

Plus I think it makes sense to have the type promotion logic closer to the kernel.

Let me know what you feel about the same (and if I missed something).

Thanks!

cc: @saketh-are

mruberry commented:
> Regarding our offline conversation about moving the type-promotion code to `where` instead of `_s_where`. I think that would be tricky. For eg. suppose `x` is `float` and `y` is `int64`. Our output should be float (which is allocated by `_s_where`. If these tensors reside on CUDA, then we don't need to cast these.

You are correct. We could give up on the CUDA optimization or pass uncast tensors if the tensor is on CUDA (which we can test for in these functions).

> So for `_s_where` to know the `dtype` of output, it will have to check which input (passed by `where`) has higher dtype and allocate output accordingly (for which we will use `promote_types`. So we are sorta doing promotion computation again)

Yes without additional refactoring the computation has to be done again to determine the output's dtype (using result_type). That's imperfect but seems OK? Alternatively if the CUDA special case is dropped then the dtype of the output can be inferred from the dtype of any input tensor. 

> Plus I think it makes sense to have the type promotion logic closer to the kernel.

I agree but all these functions seem pretty close to the kernel.


What are your thoughts on giving up on fusing the CUDA copy or special-casing CUDA inputs?

kshitij12345 committed:
move type promotion logic to where

kshitij12345 committed:
update test

kshitij12345 committed:
update scalar scalar case and remove wrapped_scalar_tensor_default_dtype

kshitij12345 committed:
update test

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

kshitij12345 commented:
Have moved the promotion logic from `_s_where` to `where`. For now we have special cased on CUDA (to recompute the computation dtype for CUDA) in `_s_where`.

This actually took care of a bug related to type promotion. (Previously, the scalar tensor was broadcasted and hence took part in type promotion which should not be the case as per https://pytorch.org/docs/stable/tensor_attributes.html#type-promotion-doc)

Have also updated the scalar-scalar case to do promotion. This allows us to get rid of the hacky code.
https://github.com/pytorch/pytorch/blob/3324bae5f106f22ff78eb6943b79cd607a1b37e4/aten/src/ATen/native/TensorCompare.cpp#L230-L250

Thanks @mruberry @saketh-are!

This should be ready for another round.

JackCaoG commented:
@kshitij12345 It seems like xla test pass, do I need to do anything on the pt/xla front?

kshitij12345 commented:
@JackCaoG the XLA tests are failing on the most recent run

https://app.circleci.com/pipelines/github/pytorch/pytorch/385123/workflows/cd4b647b-afdc-4901-8c75-ef529403f39e/jobs/16197732

JackCaoG commented:
@kshitij12345 I see, I will take care of it.

mruberry commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -318,35 +277,67 @@ static void isin_sorting(
   return at::_s_where(*b_condition, *b_self, *b_other);
 }
 
+Tensor get_promoted_tensor(const Tensor& t, ScalarType common_dtype) {
```
Add a comment explaining what this function is doing and how it's intended to be used and let's make the name specific to where ("get_promoted_where_tensor"?)

JackCaoG replied:
This will be helpful. I found it a bit hard to understand the promotion logic here. I would think we always promote both inputs to common_type but that doesn't seems to be what happens here.

mruberry commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -318,35 +277,67 @@ static void isin_sorting(
+
 Tensor where(const Tensor& condition, const Scalar& self, const Tensor& other) {
-  return at::where(condition, wrapped_scalar_tensor(self, other.device()), other);
+  auto common_dtype = at::result_type(self, other);
```
Let's add a where type promotion note and reference it in places like this where the type promotion happens so readers can more easily understand what's going on

mruberry commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -355,15 +346,18 @@ std::vector<Tensor> where(const Tensor& condition) {
 Tensor _s_where(const Tensor& condition, const Tensor& self, const Tensor& other) {
-  TORCH_CHECK(self.dtype() == other.dtype(), "expected scalar type ", self.dtype(), " but found ", other.dtype());
-  Tensor ret = at::empty(self.sizes(), self.options());
+  auto result_type = self.device().is_cpu()
```
This is going to need a comment (possibly a pointer to the where type promotion note referenced above).

Could this check be simplified to testing if self and other have distinct dtypes and, if they do, asserting they're not on the CPU?

mruberry commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4549,7 +4549,7 @@
 
 - func: where.ScalarOther(Tensor condition, Tensor self, Scalar other) -> Tensor
-  variants: function
+  variants: function, method
```
Let's not extend this to be a method in this PR

kshitij12345 replied:
Without this the newly added `cases_scalar` won't work for the method variant (for OpInfo tests). Also without this the function seems half baked.

```python
x.where(cond, y) # works
x.where(cond, 1) # doesn't work
```

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5461,10 +5461,17 @@ def random_index(shape):
              ((M, 1, M), (), (M, M, 1), True),
              ((), (M, M), (), True),)
 
+    cases_scalar = (((S, S), (S, S), 0.5, False),
```
At least one of these cases should probably be scalar x tensor and scalar x scalar, right?

Can we include a complex scalar here, too?

kshitij12345 replied:
Currently `input` has to be Tensor or a list of Tensor, thus we are checking only tensor x scalar case.

Other cases are tested in `test_torch`

mruberry replied:
Isn't the "input" tensor the boolean condition tensor, though?

kshitij12345 replied:
https://github.com/pytorch/pytorch/blob/4176afc4a035c0e79281b4955772c8962bc3ff4f/torch/testing/_internal/common_methods_invocations.py#L9651-L9655

mruberry replied:
Aha, tricky

mruberry commented on aten/src/ATen/native/TensorCompare.cpp:
```diff
@@ -318,35 +277,67 @@ static void isin_sorting(
 Tensor where(const Tensor& condition, const Scalar& self, const Tensor& other) {
-  return at::where(condition, wrapped_scalar_tensor(self, other.device()), other);
+  auto common_dtype = at::result_type(self, other);
+  const Tensor& self_t = at::scalar_tensor(self, other.options().dtype(common_dtype));
```
This PR should be sure to explain why scalar_tensor is called

https://github.com/pytorch/pytorch/blob/c0eb266c02c2886ada83662fad1107c8b5f23e1d/aten/src/ATen/native/TensorFactories.cpp#L618

as opposed to wrapped_scalar_tensor

https://github.com/pytorch/pytorch/blob/6c65ce8ee14faec577582ecd0e17ba436f32c8b3/aten/src/ATen/ScalarOps.h#L55

This assumes that it's OK to lose the information that the tensor is actually a wrapped number so long as where handles it own type promotion, right?

mruberry reviewed (commented):
Overall looks really good to me, @kshitij12345. I made a few inline comments for your review. Looking forward to an update!

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

kshitij12345 committed:
address review : add comments

kshitij12345 commented:
Gentle ping @saketh-are @mruberry

saketh-are reviewed (commented):
Hey @kshitij12345, thanks for the updates on this non-trivial PR! I left a round of comments inline, mainly around ways we might be able to cut down on the complexity of the unit tests for `torch.where`. Look forward to seeing your thoughts on this.

kshitij12345 committed:
remove irrelevant test

kshitij12345 committed:
simplify tests

kshitij12345 committed:
add function name in error

kshitij12345 committed:
simplify tests

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

kshitij12345 committed:
Add comments

kshitij12345 commented:
@saketh-are have mostly addressed all the reviews. I didn't exactly understand a few comments regarding test but have simplified the test reasonably (thanks for pointing that out :) ).

This should be ready for another round. Thanks!

saketh-are reviewed (commented):
Hey @kshitij12345, thanks for making this round of changes. The new scalar-scalar tests look really clean, plus checking against `numpy.where` is always great for confidence in the correctness of our implementation. I'm wondering if we could extend the lists of scalars you made with some handmade tensors and cover the scalar-tensor case without much additional work there? Look forward to hearing your thoughts.

kshitij12345 committed:
undo auto formatting changes

kshitij12345 committed:
address review

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

kshitij12345 committed:
address review

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

saketh-are reviewed (commented):
Hey @kshitij12345, thanks for updating `test_where_scalar_scalar`! This is looking really good. Do you think we could eliminate `test_where_scalar` before we go ahead and merge this?

kshitij12345 committed:
address review : remove test_where_scalar

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

kshitij12345 committed:
don't use mul as it type promotes

saketh-are commented:
Thanks @kshitij12345 and @JackCaoG, please feel free to ping me once the XLA fix is ready.

JackCaoG commented:
@saketh-are @kshitij12345 It seems like this fix also fix xla issue. I verified it with a small test and xla test seems to be passing on this pass. I think no change is required on the xla side.

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

kshitij12345 commented:
Thanks! @JackCaoG 

@saketh-are PTAL :)

kshitij12345 commented:
@JackCaoG I think this will need a fix at XLA side. 

The following reference test is failing : https://app.circleci.com/pipelines/github/pytorch/pytorch/403497/workflows/73e0bce0-bd07-4e77-860b-a5354ee5ff95/jobs/16643456?invite=true#step-107-3341

Also, the test from test_torch which used a fail previously is now merged into a test which is run only on Native Devices and hence get's skipped.

Depending on how complicated the fix is on XLA, we can either skip the tests on XLA for now or fix and unskip them. 

Thanks again!

kshitij12345 commented:
@JackCaoG gentle ping :)

JackCaoG commented:
Hey @kshitij12345 Please check https://github.com/pytorch/pytorch/pull/62084#issuecomment-957064183. I think no change is required from the pt/xla side.

JackCaoG commented:
@kshitij12345 Sorry, I didn't see your last comment. Let me check `test_reference_eager_where_xla_int64`

JackCaoG commented:
FYI submitted https://github.com/pytorch/xla/pull/3208, waiting for review.

JackCaoG commented:
@kshitij12345 PyTorch/XLA pr is ready, I can merge it once this pr is merged

kshitij12345 commented:
Thanks @JackCaoG 
@saketh-are XLA fix is ready. This PR should be ready for merge now. Thanks!

Commit:
remove unused variable

vadimkantorov commented:
Hope ONNX export paths are sufficiently tested for these new overloads

kshitij12345 commented:
> Hope ONNX export paths are sufficiently tested for these new overloads

Thanks for pointing it out. I am not really sure of ONNX tests. Will have a look.

Any pointer will be helpful.

cc: @saketh-are @mruberry

rgommers commented:
@saketh-are was there a problem landing this? @kshitij12345 can we give this another push and try to get it into `1.11.0`? There are multiple issues about this, and it also should address support for array API standard semantics.

kshitij12345 committed:
Merge branch 'master' into fix/where/scalar-support

kshitij12345 committed:
Merge branch 'fix/where/scalar-support' into fix/where/scalar-support

kshitij12345 committed:
make flake8 happy

kshitij12345 commented:
Gentle ping @saketh-are @mruberry

Commit:
Merge remote-tracking branch 'origin/viable/strict' into HEAD

Commit:
re-remove old tests revived during merge

JackCaoG commented:
@saketh-are Could we block this pr from merging for maybe 1 day? It has been too long and downstream xla pr has a bunch of merge conflicts and I am fixing that. I will circle back when I fix the conflict and verify the fix still work.

vadimkantorov commented:
Cool that this is getting merged! Would also be nice to have similar type promotion for `torch.lerp` (convenient for blending images): https://github.com/pytorch/pytorch/issues/57947, https://github.com/pytorch/pytorch/issues/54389

JackCaoG commented:
@saketh-are XLA pr is ready, feel free to merge this one.

kshitij12345 committed:
Merge branch 'pytorch:master' into fix/where/scalar-support

kshitij12345 commented:
@saketh-are Gentle ping :)

kshitij12345 commented:
@saketh-are ping :)

kshitij12345 commented:
@saketh-are @mruberry Ping :) Let me know if I should resolve the conflicts

vadimkantorov commented:
@ngimel Is this superseded by https://github.com/pytorch/pytorch/pull/76691 ?

ngimel commented:
yes

vadimkantorov commented:
just hope all the test cases from #9190 are added
